source/server/filter_chain_manager_impl.h

```diff
@@ -132,6 +201,8 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
   // and application protocols, using structures defined above.
   DestinationPortsMap destination_ports_map_;
   const Network::Address::InstanceConstSharedPtr address_;
+  friend class ImmediateAppendFilterChainContextCallbackImpl;
```

Comment:
Something seems off here, why do you need this to be a friend class and how do we avoid circular references with the stored shared pointers?

Reply:
@mattklein123 Sorry I don't get your question about the share_ptr par. I don't see a circular ref.

This concrete callback object is ephemeral: it doesn't really own contexts but it simplify how listener create FilterChainContext. 

The callback is a facade of FilterChainManagerImpl and it should be friend class.

Reply:
> return a shared_ptr which we assume is held elsewhere?

I lost the intention of returning shared_ptr... Probably at that time the callback/creator is not under this structure. Will change to return reference.

> When is it removed from?
In this PR the contexts are destroyed with the ListenerImpl-FilterChainManangerImpl. It's 1:1 mapping until the smart listener is implemented